lib/participant_profile_deduplicator.rb

```diff
@@ -0,0 +1,150 @@
+# frozen_string_literal: true
+
+class ParticipantProfileDeduplicator
+  CONFLICTING_DECLARATION_STATES = %w[submitted eligible payable paid].freeze
+
+  attr_reader :primary_profile_id, :duplicate_profile_id, :dry_run
+
+  def initialize(primary_profile_id, duplicate_profile_id, dry_run: true)
+    @primary_profile_id = primary_profile_id
+    @duplicate_profile_id = duplicate_profile_id
+    @dry_run = dry_run
+  end
+
+  def dedup!
+    Rails.logger.info("~~~ DRY RUN ~~~") if dry_run
+
+    ensure_lead_providers_match
+
+    ActiveRecord::Base.transaction do
+      handle_school_change!
+      transfer_validation_data!
+      transfer_participant_eligibility!
+      reconcile_declarations!
+      delete_duplicate!
+
+      raise ActiveRecord::Rollback if dry_run
+    end
+  end
+
+private
+
+  def ensure_lead_providers_match
+    return if primary_profile.lead_provider.id == duplicate_profile.lead_provider.id
+
+    raise "Only duplicates with the same lead_provider are supported"
+  end
+
+  def handle_school_change!
+    return unless primary_profile.school != duplicate_profile.school
+
+    raise "Only duplicates with a single induction record are supported" unless duplicate_profile.induction_records.one?
+
+    duplicate_induction_record = duplicate_profile.latest_induction_record
+    end_date = primary_profile.induction_records.oldest.start_date - 1.minute
+
+    duplicate_induction_record.update!(
+      participant_profile_id: primary_profile.id,
+      induction_status: :leaving,
+      end_date:,
+      preferred_identity: preferred_identity(duplicate_induction_record),
+    )
+
+    log_info("Preferred identity updated on duplicate induction record") if duplicate_induction_record.saved_change_to_preferred_identity_id?
+
+    log_info("Duplicate profile induction record transferred. End date: #{end_date}")
+
+    primary_profile.latest_induction_record.update!(school_transfer: true)
```

Comment:
I think we need to set only the oldest induction record coming after the transfer to `school_transfer` true